### PR TITLE
Splits code to seperate modules & add documention

### PR DESCRIPTION
**Pull request by aopincar**

### CODE REVIEW

File: tools/kcli/kcli/parse.py [hunk 1]
```diff
@@ -54,9 +54,9 @@ def create_parser(options_trees):
         sub_parser.add_argument("-e", "--extra-vars", default=list(),
                                 action='append', help="Provide extra vars")
         sub_parser.add_argument("-n", "--input", action='append',
-                                help="a settings file that will be loaded "
-                                     "first, all other settings file will be"
-                                     " merged with it", default=list())
+                                help="Settings files to be loaded first,"
+                                     " other settings file will be"
```
Comment (yfried) [inline tools/kcli/kcli/parse.py]: s/file/files

File: tools/kcli/kcli/parse.py [hunk 2]
```diff
@@ -54,9 +54,9 @@ def create_parser(options_trees):
...
+                                     " merged with them", default=list())
```
Comment (yfried) [inline tools/kcli/kcli/parse.py:62]: What's the merge order/priority? will "input" files override the "options" or vice-versa? what about "extra-vars"?

File: tools/kcli/kcli/options.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
+import os
```
Comment (yfried) [inline tools/kcli/kcli/options.py:7]: Please add a short docstring describing this module

File: tools/kcli/kcli/options.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+import yaml
+
+from kcli import conf
+from kcli import exceptions
+from kcli import logger
+
+LOG = logger.LOG
+
+
+class OptionNode(object):
```
Comment (yfried) [inline tools/kcli/kcli/options.py:18]: missing class doc

File: tools/kcli/kcli/options.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            self.option = "-".join([self.parent.option, self.option])
+        self.values = self._get_values()
+        self.children = {i: dict() for i in self._get_sub_options()}
+
+        if self.parent:
+            self.parent_value = self.path.split("/")[-2]
+            self.parent.children[self.parent_value][self.option] = self
+
+    def _get_values(self):
+        """
```
Comment (yfried) [inline tools/kcli/kcli/options.py]: on-line docstring should be:

``` python

"""Returns a sorted list of values available for the current option. """
```

File: tools/kcli/kcli/options.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        options = [options_dir for options_dir in os.listdir(self.path)
+                   if os.path.isdir(os.path.join(self.path, options_dir)) and
+                   options_dir in self.values]
+
+        options.sort()
+        return options
+
+
+class OptionsTree(object):
+    def __init__(self, settings_dir, option):
```
Comment (yfried) [inline tools/kcli/kcli/options.py:67]: missing class doc. what is "options tree"?

File: tools/kcli/kcli/options.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        :return: list of paths to settings files of 'options'
+        """
+        ymls = []
+        if not options:
+            return ymls
+
+        keys = options.keys()
+        keys.sort()
+
+        def step_in(key, node):
```
Comment (yfried) [inline tools/kcli/kcli/options.py:136]: short doc?

File: tools/kcli/kcli/utils.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
+import os
```
Comment (yfried) [inline tools/kcli/kcli/utils.py:5]: Please add a short docstring describing this module

File: tools/kcli/kcli/utils.py [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+
+
+# TODO: remove "settings" references in project
+def validate_settings_dir(settings_dir=None):
+    """Checks & returns the full path to the settings dir.
+
+    Path is set in the following priority:
+    1. Method argument
+    2. System environment variable
+    3. Settings dir in the current working dir
```
Comment (yfried) [inline tools/kcli/kcli/utils.py]: I don't think this should be an option. we shouldn't rely on current working dir for this.

File: tools/kcli/kcli/utils.py [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+    :param settings_dir: path given as argument by a user
+    :return: path to settings dir (str)
+    :raise: IRFileNotFoundException: when the path to the settings dir doesn't
+            exist
+    """
+    settings_dir = settings_dir or os.environ.get(
+        'KHALEESI_SETTINGS') or os.path.join(os.getcwd(), "settings", "")
```
Comment (yfried) [inline tools/kcli/kcli/utils.py]: Please define "KHALEESI_SETTINGS" in conf.py as a constant

File: tools/kcli/kcli/utils.py [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+            settings_dir,
+            "Settings dir doesn't exist: ")
+
+    return settings_dir
+
+
+def merge_settings(settings, file_path):
+    """merge settings in 'file_path' with 'settings'
+
+    :param settings: settings to be merge with
```
Comment (yfried) [inline tools/kcli/kcli/utils.py]: what object is this? dict? OptionsTree?

File: tools/kcli/kcli/utils.py [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+
+    if not os.path.exists(settings_dir):
+        raise exceptions.IRFileNotFoundException(
+            settings_dir,
+            "Settings dir doesn't exist: ")
+
+    return settings_dir
+
+
+def merge_settings(settings, file_path):
```
Comment (yfried) [inline tools/kcli/kcli/utils.py]: This method is a bit weird. it merges some "settings" object (dict? optionsTree?) with a file.
The name and doc should be more about "update 'settings object' with file"

File: tools/kcli/kcli/utils.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+        raise exceptions.IRFileNotFoundException(file_path)
+
+    loaded_file = configure.Configuration.from_file(file_path).configure()
+    settings = settings.merge(loaded_file)
+
+    return settings
+
+
+def generate_settings_file(settings_files, extra_vars):
+    """Generate one settings file from a given list of settings files and
```
Comment (yfried) [inline tools/kcli/kcli/utils.py]: - it doesn't generate a file. it merges all input into a single "settings dict".
- please specify the merge priority. Example: "files from list, then vars. last one wins"

File: tools/kcli/kcli/utils.py [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+    :return: dictionary with merging results of all settings file and
+    extra-vars
+    """
+    settings = configure.Configuration.from_dict({})
+
+    for settings_file in settings_files:
+        settings = merge_settings(settings, settings_file)
+
+    for extra_var in extra_vars:
+        if extra_var.startswith('@'):
```
Comment (yfried) [inline tools/kcli/kcli/utils.py:142]: `extra-vars` accepts files as well? Is this doc'ed in "help"? How is it different than the files in "input"?

File: tools/kcli/kcli/utils.py [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+
+    example:
+      dict_lookup({}, 'val', 'key1.key2'.split('.'))
+
+    :param dic: a dictionary object to insert the nested key value into
+    :param val: a value to insert to the given dictionary
+    :param key: first key in a chain of key that will store the value
+    :param keys: sub keys in the keys chain
+    """
+    if not keys:
```
Comment (yfried) [inline tools/kcli/kcli/utils.py:76]: how about:

``` python
keys.insert(0, key)
```

File: tools/kcli/kcli/utils.py [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+        dic[key] = val
+        return
+
+    if key not in dic:
```
Comment (yfried) [inline tools/kcli/kcli/utils.py]: better to use `setdefault`

File: tools/kcli/kcli/utils.py [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+        dic[key] = {}
+
+    dict_insert(dic[key], val, *keys)
```
Comment (yfried) [inline tools/kcli/kcli/utils.py]: Please consider this alternative:

``` python

def dict_recurs(dic, val, list_of_keys):
    if len(list_of_keys):
        key = list_of_keys.pop(0)
        dic[key] = val
        dict_recurs(dic, val, list_of_keys)
    else:
         return

```

File: tools/kcli/kcli/utils.py [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+    # Dump & load again settings, because 'in_string_lookup' can't work with
+    # 'Configuration' object.
+    dumped_settings = yaml.safe_dump(settings, default_flow_style=False)
+    settings = yaml.safe_load(dumped_settings)
+
+    return settings
+
+
+def in_string_lookup(settings):
+    """convert strings contain the '!lookup' tag in them and don't
```
Comment (yfried) [inline tools/kcli/kcli/utils.py:161]: please explain in doc how this can happen (nested lookup?)

File: tools/kcli/kcli/yamls.py [hunk 19]
```diff
@@ -1,13 +1,22 @@
+import re
```
Comment (yfried) [inline tools/kcli/kcli/yamls.py:5]: Please add short doc describing the module

File: tools/kcli/kcli/parse.py (hunk 2 above)
Comment (aopincar) [inline tools/kcli/kcli/parse.py:62]: nope.
input files are the first to be loaded, after them the "options" settings files will be loaded and in the end are the extra-vars.
I agree this should be documented but not sure if in the help message or in the README file.
What do you think?

File: tools/kcli/kcli/utils.py (hunk 15 above)
Comment (aopincar) [inline tools/kcli/kcli/utils.py:76]: why? I want to explicitly set the value of key and exit - if there's only one key

File: tools/kcli/kcli/utils.py (hunk 16 above)
Comment (aopincar) [inline tools/kcli/kcli/utils.py]: +1

File: tools/kcli/kcli/utils.py (hunk 14 above)
Comment (aopincar) [inline tools/kcli/kcli/utils.py:142]: answered the priority in upper comment.
the extra vars indeed accept files, for this it file path should start with "@".
Here also I think it should be mentioned in the README file.

File: tools/kcli/kcli/parse.py (hunk 2 above)
Comment (aopincar) [inline tools/kcli/kcli/parse.py:62]: I had mentioned the merging order in the README.rst file ("Merging order")

File: tools/kcli/kcli/parse.py [hunk 20]
```diff
@@ -54,9 +54,9 @@ def create_parser(options_trees):
         sub_parser.add_argument("-e", "--extra-vars", default=list(),
                                 action='append', help="Provide extra vars")
```
Comment (yfried) [inline tools/kcli/kcli/parse.py]: Please address the "@" syntax

File: tools/kcli/README.rst [hunk 21]
```diff
@@ -54,4 +54,23 @@ This displays options you can pass to ``kcli``.
 
   $ export WORKSAPCE=$(dirname `pwd`)
 
-
+Extra-Vars
+----------
+One can set/overwrite settings in the output file using the '-e/--extra-vars'
+option. There are 2 ways of doing so:
+1. specific settings: (key=value form)
```
Comment (yfried) [inline tools/kcli/README.rst:62]: needs newline for numbers to work.

File: tools/kcli/README.rst [hunk 22]
```diff
@@ -54,4 +54,23 @@ This displays options you can pass to ``kcli``.
...
+
+The '-e/--extra-vars' can be used more than once.
+
+Merging order
+-------------
+Except options based on the settings dir structure, kcli accepts input of
+predefined settings files (with -n/--input) and user defined specific options
+(-e/--extra-vars).
+The merging priority order listed below:
+1. Input files
```
Comment (yfried) [inline tools/kcli/README.rst:76]: needs newline for numbers to work.

**Pull request merged**